spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java

```diff
@@ -315,16 +317,37 @@ private boolean obtainLock() {
 			 */
 			this.lockedAt = System.currentTimeMillis();
 			this.threadName = currentThread.getName();
-			Boolean success = RedisLockRegistry.this.redisTemplate.boundValueOps(
-					constructLockKey()).setIfAbsent(this);
+
+			Boolean success = RedisLockRegistry.this.redisTemplate.execute(new SessionCallback<Boolean>() {
+				@SuppressWarnings("unchecked")
+				@Override
+				public Boolean execute(RedisOperations ops) throws DataAccessException {
+					String key = constructLockKey();
+
+					ops.watch(key); //monitor key
+
+					ops.multi(); //transaction start
+
+					//can't relay on operations result inside transaction, execution is delayed till `exec()`
```

Comment:
I'm not well with English, so I'm going to remove these comments during merge :smile: 
Anyway let's discuss what you mean, since your name is Russian I hope you talk that language too.
`Can't rely` - `Не можем полагаться`. True ?